docs/source/triggers/index.rst

```diff
@@ -125,159 +111,316 @@ Like ``Scenario`` or ``Update``, you can use ``do`` notation and
+choice, as it now needs to set the ``receivedAt`` field. Here is the updated
+code for ``SendMessage``:
+
+.. code-block:: daml
+
+    -- New definition for SendMessage
+        nonconsuming choice SendMessage: ContractId Message with
+            sender: Party
+            content: Text
+          controller sender
+          do
+            assertMsg "Designated user must follow you back to send a message" (elem sender following)
+            now <- getTime
+            create Message with sender, receiver = username, content, receivedAt = now
+
+The ``getTime`` action (`doc </daml/stdlib/Prelude.html#function-da-internal-lf-gettime-99334>`__)
+returns the time at which the command was received by the sandbox. In more
+sensitive applications, this may not be sufficiently reliable, as transactions
+may be processed in parallel (so "received at" timestamp order may not match
+actual transaction order), and in distributed cases dishonest participants may
+fudge this value. It's good enough for this example, though.
+
+Now that we have a field to sort on, and thus a way to identify the *latest*
+message, we can turn our attention back to our trigger code.
+
+AutoReply
+=========
+
+Open up the trigger code again (``daml/ChatBot.daml``), and change it to:
+
+.. code-block:: daml
+
+    module ChatBot where
+
+    import qualified Daml.Trigger as T
+    import qualified User
+    import qualified DA.List.Total as List
+
+    autoReply : T.Trigger ()
+    autoReply = T.Trigger
+      { initialize = pure ()
+      , updateState = \_ -> pure ()
+      , rule = \p -> do
+          message_contracts <- T.query @User.Message
+          let messages = map snd message_contracts
+          debug $ "Messages so far: " <> show (length messages)
+          let lastMessage = List.last $ List.sortOn (.receivedAt) messages
+          debug $ "Last message: " <> show lastMessage
+          case lastMessage of
+            Some m ->
+              if m.receiver == p
+              then do
+                users <- T.query @User.User
+                debug users
+                let isSender = (\user -> user.username == m.sender)
+                let replyTo = List.head $ filter (\(_, user) -> isSender user) users
+                case replyTo of
+                  None -> pure ()
+                  Some (sender, _) -> do
+                    T.dedupExercise sender (User.SendMessage p "Please, tell me more about that.")
+              else pure ()
+            None -> pure ()
+      , registeredTemplates = T.AllInDar
+      , heartbeat = None
+      }
+
+Refresh ``daml start`` by pressing ``r`` (followed by ``Enter`` on Windows) in
+its terminal, then kill (CTRL-C) the trigger and start it again. Play a bit
+with ``alice`` and ``bob`` in your browser, to get a feel for how the
+trigger works. Watch both the messages in-browser and the debug statements
+printed by the trigger runner.
+
+Let's walk through the ``rule`` code line-by-line:
```

Comment:
The explanation of trigger behavior that makes the most sense to me is to view them as trying to go change the current ACS to a target ACS and then they become a noop. The event handling descriptions usually end up being a bit fuzzy and confusing imho and lead to wrong expectations around when the trigger gets triggered.

In your case, the trigger tries to change the ACS such that the last message was sent by it and then stops.

Reply:
I've added a section at the top, and tried to tweak the step-by-step explanations to go in that direction.